projects/httpProxy.c

```diff
@@ -0,0 +1,164 @@
+        exit(EXIT_FAILURE);
+    }
+
+    printf("HTTP Proxy Server Listening on port %d\n", PORT);
+
+    while (1) {
+        // Принятие соединения от клиента
+        if ((client_socket = accept(server_fd, (struct sockaddr*)&address, (socklen_t*)&addr_len)) < 0) {
+            perror("accept");
+            continue;
+        }
+
+        // Создание потока для обработки соединения
+        pthread_t thread_id;
+        if (pthread_create(&thread_id, NULL, handle_connection, (void*)&client_socket) != 0) {
+            perror("pthread_create");
+            continue;
+        }
+    }
+
+    return 0;
+}
+
+void* handle_connection(void* client_socket) {
+    int client_sock = *(int*)client_socket;
+    char buffer[BUFFER_SIZE];
+    int target_sock, read_bytes;
+
+    // Чтение запроса от клиента
+    read_bytes = recv(client_sock, buffer, BUFFER_SIZE - 1, 0);
+    if (read_bytes <= 0) {
+        close(client_sock);
+        return NULL;
+    }
+    buffer[read_bytes] = '\0'; // Добавляем нуль-терминатор
+
+    // Извлечение имени хоста из запроса
+    char host[_SC_HOST_NAME_MAX + 1] = {0};
+    char* host_line_start = strstr(buffer, "Host: ");
+    if (!host_line_start) {
+        host_line_start = strstr(buffer, "host: "); // Попытка найти заголовок в другом регистре
+    }
+
+    if (host_line_start) {
+        char* host_line_end = strstr(host_line_start, "\r\n");
+        if (!host_line_end) {
+            host_line_end = strstr(host_line_start, "\n");
+        }
+
+        if (host_line_end) {
+            int host_length = host_line_end - (host_line_start + 6);
+            strncpy(host, host_line_start + 6, host_length);
+            host[host_length] = '\0'; // Обеспечиваем правильное завершение строки
+        } else {
+            printf("Host header line end not found.\n");
+        }
+    } else {
+        printf("Host header not found in request.\n");
+        close(client_sock);
+        return NULL;
+    }
+
+    printf("Connecting to host: %s\n", host);
+
+    // Подключение к целевому серверу
+    target_sock = connect_to_target(host, PORT); // Используем PORT, определенный в начале программы
+    if (target_sock < 0) {
+        close(client_sock);
+        return NULL;
+    }
+
+    // Передача запроса на целевой сервер
+    send(target_sock, buffer, strlen(buffer), 0);
+
+    // Передача ответа обратно клиенту
+    while ((read_bytes = recv(target_sock, buffer, BUFFER_SIZE, 0)) > 0) {
+        send(client_sock, buffer, read_bytes, 0);
+    }
+
+    // Закрытие соединений
+    close(target_sock);
+    close(client_sock);
+    return NULL;
+}
+
+int connect_to_target(char* host, int port) {
+    struct hostent *he;
+    struct sockaddr_in server_addr;
+    int sock;
+
+    // Получение информации о хосте
+    if ((he = gethostbyname(host)) == NULL) {
+        herror("gethostbyname failed");
+        return -1;
+    }
+
+    // Создание сокета
+    if ((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
```

Comment:
Please list all possible socket doimains, protocol and types with brief description

Reply:
Domains
AF_UNIX, AF_LOCAL -> Local communication
AF_INET -> IPv4 Internet protocols
AF_INET6 -> IPv6 Internet protocols
AF_IPX -> IPX - Novell protocols
AF_NETLINK -> Kernel user interface device
AF_X25 -> ITU-T X.25 / ISO-8208 protocol
AF_AX25 -> Amateur radio AX.25 protocol
AF_ATMPVC -> Access to raw ATM PVCs
AF_APPLETALK -> Appletalk
AF_PACKET -> Low level packet interface

socket types:
SOCK_STREAM Provides sequenced, reliable, two-way, connection-based byte streams. An out-of-band data transmission mechanism may be supported.
SOCK_DGRAM Supports datagrams (connectionless, unreliable messages of a fixed maximum length).
SOCK_SEQPACKET Provides a sequenced, reliable, two-way connection-based data transmission path for datagrams of fixed maximum length; a consumer is required to read an entire packet with each input system call.
SOCK_RAW Provides raw network protocol access.
SOCK_RDM Provides a reliable datagram layer that does not guarantee ordering.
SOCK_PACKET Obsolete and should not be used in new programs.
SOCK_NONBLOCK Set the O_NONBLOCK file status flag on the new open file description.
SOCK_CLOEXEC Set the close-on-exec (FD_CLOEXEC) flag on the new file descriptor.